packages/@lwc/ssr-compiler/src/compile-template/shared.ts

```diff
@@ -25,41 +27,108 @@ import type {
     ObjectExpression as EsObjectExpression,
     Property as EsProperty,
     Statement as EsStatement,
+    IfStatement as EsIfStatement,
+    YieldExpression as EsYieldExpression,
+    ExpressionStatement as EsExpressionStatement,
 } from 'estree';
 import type {
     ComplexExpression as IrComplexExpression,
     Expression as IrExpression,
     Literal as IrLiteral,
 } from '@lwc/template-compiler';
 
+const bYieldTernary =
+    esTemplateWithYield`yield ${is.expression} ? ${is.expression} : ${is.expression}`<EsExpressionStatement>;
+
+interface OptimizableYield extends EsExpressionStatement {
+    expression: EsYieldExpression & { delegate: false };
+}
+
+const bOptimizedYield = esTemplateWithYield`yield ${is.expression};`<OptimizableYield>;
+
+function isOptimizableYield(stmt: EsStatement | undefined): stmt is OptimizableYield {
+    return (
+        is.expressionStatement(stmt) &&
+        is.yieldExpression(stmt.expression) &&
+        stmt.expression.delegate === false
+    );
+}
+
+/** Returns null if the statement cannot be optimized. */
+function optimizeSingleStatement(stmt: EsStatement): OptimizableYield | null {
+    if (is.blockStatement(stmt)) {
+        // `if (cond) { ... }` => optimize inner yields and see if we can condense
+        const optimizedBlock = optimizeAdjacentYieldStmts(stmt.body);
+        // More than one statement cannot be optimized into a single yield
+        if (optimizedBlock.length !== 1) return null;
+        const [optimized] = optimizedBlock;
+        return isOptimizableYield(optimized) ? optimized : null;
+    } else if (is.expressionStatement(stmt)) {
+        // `if (cond) expression` => just check if expression is a yield
+        return is.yieldExpression(stmt) ? stmt : null;
+    } else {
+        // Can only optimize expression/block statements
+        return null;
+    }
+}
+
+/**
+ * Tries to reduce if statements that only contain yields into a single yielded ternary
+ * Returns null if the statement cannot be optimized.
+ */
+function optimizeIfStatement(stmt: EsIfStatement): EsExpressionStatement | null {
+    const consequent = optimizeSingleStatement(stmt.consequent)?.expression.argument;
+    if (!consequent) {
+        return null;
+    }
+
+    const alternate = stmt.alternate
+        ? optimizeSingleStatement(stmt.alternate)?.expression.argument
+        : b.literal('');
+    if (!alternate) {
+        return null;
+    }
+
+    return bYieldTernary(stmt.test, consequent, alternate);
+}
+
 export function optimizeAdjacentYieldStmts(statements: EsStatement[]): EsStatement[] {
```

Comment:
@wjhsf do you feel we have enought test coverage to be sure we're covering all the scenarios before/after this change?

Reply:
👍 🚢